R/init.R

```diff
@@ -385,6 +395,54 @@ if (interactive() &&
         page_viewer = page_viewer
       )
 
+      #vsc-r-api
+      RESPONSE_TIMEOUT <- 30
+      response_lock_file <- file.path(dir_extension, "response.lock")
+      response_file <- file.path(dir_extension, "response.log")
+      addin_registry <- file.path(dir_session, "addins.json")
+      
+      get_response_timestamp <- function() {
+        if (file.exists(response_lock_file))
+          readLines(response_lock_file)
+        else NA
+      }
+      # initialise the reponse timestamp to NA or the value currently in the file
+      response_time_stamp <- get_response_timestamp()
+
+      get_response_lock <- function() {
+        lock_time_stamp <- get_response_timestamp()
+        if (isTRUE(lock_time_stamp != response_time_stamp)) {
+          response_time_stamp <<- lock_time_stamp
+          TRUE
+        } else FALSE
+      }
+
+      request_response <- function(command, ...) {
+        request(command, ...)
+        wait_start <- Sys.time()
+        while (!get_response_lock()) {
+          if ((Sys.time() - wait_start) > RESPONSE_TIMEOUT) 
+            stop("Did not receive a response from VSCode-R API within ", 
+                  RESPONSE_TIMEOUT, " seconds.")
+          Sys.sleep(0.1)
+        }
+        jsonlite::read_json(response_file)
+      }
+
+     source(file.path(dir_extension, "vsc_r_api_util.R"),
+       local = TRUE
+     )
+     source(file.path(dir_extension, "vsc_r_api.R"),
+       local = TRUE
+     )
+     update_addin_registry(addin_registry)
```

Comment:
Maybe we should call this in `attach()` function in `.vsc`? Then each attach call refreshes the addins so that user could re-attach after installing a new addin to get vscode-R noticed.

Reply:
Yes and when I refactor this, I must:
* reset the 'warmup' back to 200
* make it so the addin picker can only run if a terminal has been attached.